File: P1_Two_Sum/Python/other_imp2/solution.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        #Two Pointers Implementation | Time: O(nlogn) | Space: O(n)
        if not nums:
            return None

        #sorting
        result = []
        for i, num in enumerate(nums):
            result.append((num, i))

        result.sort(key=lambda x:x[0])

        #two pointers
        i, j = 0, len(result)-1
        while i<j: #overlap of i and j is not allowed since our mission is to find a 'pair'
            if result[i][0] + result[j][0] == target:
                return [result[i][1], result[j][1]] #correct pair found
            elif result[i][0] + result[j][0] < target:# too small, ++i
                i += 1
            else: #too large j--
                j -= 1

        return None

File: P1_Two_Sum/CPlusPlus/other_imp2/solution.cpp
#include <map>
#include <iostream>
#include <vector>
#include <cstdio>
#include <cstdlib>
#include <algorithm>
#include <solution.h>

/*
std::vector<int> Solution::twoSum(std::vector<int>& nums, int target) {
    std::map<int, int> hash_map;
    std::vector<int>::iterator vec_begin = nums.begin();
    std::vector<int>::iterator vec_end = nums.end();

    for (auto it = vec_begin; it != vec_end; ++it){
        int complement = target - *it;

        if (hash_map.find(complement)!=hash_map.end()) {
            std::vector<int> answer_vec (2, 0);
            answer_vec[0] = hash_map.find(complement)->second;
            answer_vec[1] = it - vec_begin;
            return answer_vec;
        }else{
            hash_map[*it] = it - vec_begin;
        }
    }

    printf("Error: No two sum solution.\n");
    exit (EXIT_FAILURE);
}
*/
std::vector<int> Solution::twoSum(std::vector<int>& nums, int target) {
    // Two Pointers Implementation | Time: O(nlogn) | Space: O(n)
    std::vector<int> answer_vec (2, 0);
    std::vector<std::pair<int, int>> result; //(num, index) tuple
    bool got_answer = false;

    //sorting
    for (size_t i=0, vec_size=nums.size(); i!=vec_size; ++i){
        std::pair<int, int> pair_idx = std::make_pair(nums[i], (int)i);
        result.push_back(pair_idx);
    }

    std::sort(result.begin(), result.end(), 
    [](const auto& a, const auto& b){return a.first < b.first;});

    //two pointers
    int i = 0;
    int j = nums.size()-1;

    while(i<j){
        if(result[i].first + result[j].first == target){
            answer_vec[0] = result[i].second;
            answer_vec[1] = result[j].second;
            got_answer = true;
            break;
        }else if(result[i].first + result[j].first < target){//too small, ++i
            ++i;
        }else{//too large, --j
            --j;
        }
    }

    if(!got_answer){
        printf("Error: No two sum solution.\n");
        exit (EXIT_FAILURE);
    }
    return answer_vec;
}
